Qt: avoid AskPassphraseDialog synchronous QDialog.exec() calls

jonasschnelli opened the pull request:
Fixes #15310

If this is an acceptable approach then this shall be adapted (eventually in this PR) to other areas where we use `QDialog.exec()`.

The synchronous `QDialog.exec()` call interacts badly with multiwallet (especially unloading wallets).

From the Qt docs:

> Note: Avoid using this function; instead, use open(). Unlike exec(), open() is asynchronous, and does not spin an additional event loop. This prevents a series of dangerous bugs from happening (e.g. deleting the dialog's parent while the dialog is open via exec()). When using open() you can connect to the finished() signal of QDialog to be notified when the dialog is closed.

jonasschnelli committed:
Qt: avoid AskPassphraseDialog synchronous QDialog.exec() calls

promag reviewed (commented):
Concept ACK.

Is `unique_ptr` necessary? Could use `deleteLater` slot instead?

nit 1, `AskPassphraseDialog` constructor could do `setModal(true)`?
nit 2, `AskPassphraseDialog` could receive wallet model instance?

flack commented:
maybe that's a stupid question, but is there any reason not to make a function which opens the dialog? Because to me it looks like it's basically three times the same block of code in the same file with only a different first argument in `AskPassphraseDialog` (ok, and the second `connect` call in the first code block, but that could also be in an `if` I guess)

(Disclaimer: Not a C++ expert)

jonasschnelli commented:
Followed @promag advice and changed...
... from `unique_ptr` to usage of QT's `deleteLater` which I think is the superior solution (the only advantage of a `unique_ptr` would be that even if `QDialog::finished` will never gets fired [bug in QT] it would deallocate the instance correctly on opening of a new dialog or unloading a `walletView`,.... also, I think the `deleteLater` approach will confuse static analysers [pretty sure they will report a leak]).
... moved the `setModal()` to the AskPassphraseDialog constructor

The by @promag proposed walletModel change should be done in another PR since its a medium size refactor.

promag commented on src/qt/askpassphrasedialog.cpp:
```diff
@@ -74,6 +74,9 @@ AskPassphraseDialog::AskPassphraseDialog(Mode _mode, QWidget *parent) :
     connect(ui->passEdit3, &QLineEdit::textChanged, this, &AskPassphraseDialog::textChanged);
+
+    /* always use modal userflow blocking */
+    setModal(true);
```
Can also `setAttribute(Qt::WA_DeleteOnClose)` and ditch connections to `deleteLater`?

jonasschnelli commented:
Thanks for the further optimisation @promag!
Implemented your suggestion.

laanwj commented on src/qt/askpassphrasedialog.cpp:
```diff
@@ -74,6 +74,12 @@ AskPassphraseDialog::AskPassphraseDialog(Mode _mode, QWidget *parent) :
+    setModal(true);
+
+    // auto-delete instance after close event
+    setAttribute(Qt::WA_DeleteOnClose);
```
I don't think this should be set in the constructor. It is not the dialog's responsibility to know about its own object lifetime. So I think it's more clear to do so at the call site.

promag replied:
This was suggested by me in https://github.com/bitcoin/bitcoin/pull/15313#discussion_r253711302 and now I agree with you. @jonasschnelli sorry ☺️

jonasschnelli replied:
Yes. I agree with @laanwj. Fixed.

laanwj commented:
Concept ACK

laanwj commented:
LGTM now utACK https://github.com/bitcoin/bitcoin/pull/15313/commits/5e8dadbcec674a09f6836075bad531395c294399

promag committed:
gui: Defer removeAndDeleteWallet when no modal widget is active

promag commented on src/qt/walletview.cpp:
```diff
@@ -241,11 +241,11 @@ void WalletView::encryptWallet(bool status)
-    dlg.exec();
-
-    updateEncryptionStatus();
+    AskPassphraseDialog *dlg = new AskPassphraseDialog(status ? AskPassphraseDialog::Encrypt : AskPassphraseDialog::Decrypt, this);
```
nit, here and below: `AskPassphraseDialog* dlg`.

promag reviewed (commented):
Crashes with the following:

```sh
./src/qt/bitcoin-qt -regtest -wallet=x1 -server
```
Then open Settings -> Encrypt Wallet..

<img width="1028" alt="Screenshot 2019-03-17 at 22 32 37" src="https://user-images.githubusercontent.com/3534524/54498941-bc569b00-4904-11e9-8440-36369e1db689.png">
Confirm and when the following dialog opens:
<img width="532" alt="Screenshot 2019-03-17 at 22 32 55" src="https://user-images.githubusercontent.com/3534524/54498943-beb8f500-4904-11e9-8499-4f7cf803d438.png">
Run

```sh
./src/bitcoin-cli -regtest unloadwallet x1
```
Should give something like:

```
bitcoin-qt(70767,0x1116265c0) malloc: *** error for object 0x7ffeecb93160: pointer being freed was not allocated
bitcoin-qt(70767,0x1116265c0) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    70767 abort      ./src/qt/bitcoin-qt -regtest -wallet=x1 -server
```

jonasschnelli commented:
As said:

> If this is an acceptable approach then this shall be adapted (eventually in this PR) to other areas where we use QDialog.exec().

There are still a couple of places where dialogs are "synchronous" and this needs to be fixed... but I agree with @promag that the one around encryption should be fixes to cover the complete issue. I'll have a look soon.

jonasschnelli committed:
Only delay walletmodel removal when modal dialogs are synchronous

jonasschnelli commented:
Combined this with #15614 to gain a safe unloading-wallet behaviour in the GUI

jonasschnelli commented:
Closing for now in favour of #15614
